src/lib/operators/join_mpsm/radix_cluster_sort_numa.hpp

```diff
@@ -0,0 +1,368 @@
+  static size_t _materialized_table_size(MaterializedNUMAPartition<T>& table) {
+    size_t total_size = 0;
+    for (auto chunk : table._chunk_columns) {
+      total_size += chunk->size();
+    }
+
+    return total_size;
+  }
+
+  /**
+  * Concatenates multiple materialized columns to a single materialized column.
+  **/
+  static std::unique_ptr<MaterializedNUMAPartitionList<T>> _concatenate_chunks(
+      std::unique_ptr<MaterializedNUMAPartitionList<T>>& input_table) {
+    // We know input chunks has only one numa_node
+    auto output_table = std::make_unique<MaterializedNUMAPartitionList<T>>();
+    output_table->push_back(MaterializedNUMAPartition<T>(NodeID{0}, 1));
+
+    // Reserve the required space and move the data to the output
+    auto output_chunk = std::make_shared<MaterializedChunk<T>>();
+    _materialized_table_size((*input_table)[0]);
+    for (auto& chunk : (*input_table)[0]._chunk_columns) {
+      output_chunk->insert(output_chunk->end(), chunk->begin(), chunk->end());
+    }
+
+    (*output_table)[0]._chunk_columns[0] = output_chunk;
+
+    return output_table;
+  }
+
+  /**
+  * Performs the clustering on a materialized table using a clustering function that determines for each
+  * value the appropriate cluster id. This is how the clustering works:
+  * -> Count for each chunk how many of its values belong in each of the clusters using histograms.
+  * -> Aggregate the per-chunk histograms to a histogram for the whole table. For each chunk it is noted where
+  *    it will be inserting values in each cluster.
+  * -> Reserve the appropriate space for each output cluster to avoid ongoing vector resizing.
+  * -> At last, each value of each chunk is moved to the appropriate cluster.
+  **/
+  MaterializedNUMAPartition<T> _cluster(MaterializedNUMAPartition<T>& input_chunks,
+                                        std::function<size_t(const T&)> clusterer, NodeID node_id) {
+    auto num_chunks = input_chunks._chunk_columns.size();
+    auto output_table = MaterializedNUMAPartition<T>(node_id, _cluster_count);
+    NUMAPartitionInformation numa_part_info(num_chunks, _cluster_count);
+
+    // Count for every chunk the number of entries for each cluster in parallel
+    for (size_t chunk_number = 0; chunk_number < num_chunks; ++chunk_number) {
+      auto& chunk_information = numa_part_info.chunk_information[chunk_number];
+      auto input_chunk = input_chunks._chunk_columns[chunk_number];
+
+      for (auto& entry : *input_chunk) {
+        auto cluster_id = clusterer(entry.value);
+        ++chunk_information.cluster_histogram[cluster_id];
+      }
+    }
+
+    // Aggregate the chunks histograms to a table histogram and initialize the insert positions for each chunk
+    for (auto& chunk_information : numa_part_info.chunk_information) {
+      for (size_t cluster_id = 0; cluster_id < _cluster_count; ++cluster_id) {
+        chunk_information.insert_position[cluster_id] = numa_part_info.cluster_histogram[cluster_id];
```

Comment:
why call it `chunk_information`, but `numa_part_info`?

Reply:
Because an acronym slipped through :D